File: 1197-parsing-a-boolean-expression/1197-parsing-a-boolean-expression.c
#include <stdbool.h>
#include <stdio.h>
#include <string.h>

bool parseBoolExpr(char* expression) {
    char stack[2000];
    int top = -1;

    for (int i = 0; expression[i] != '\0'; i++) {
        char ch = expression[i];

        if (ch == ',' || ch == '(') {
            continue;
        }

        if (ch == 't' || ch == 'f' || ch == '!' || ch == '&' || ch == '|') {
            stack[++top] = ch;
        } else if (ch == ')') {
            bool hasTrue = false, hasFalse = false;
            while (stack[top] == 't' || stack[top] == 'f') {
                if (stack[top] == 't') {
                    hasTrue = true;
                } else {
                    hasFalse = true;
                }
                top--; // Pop the current value
            }

            char operator= stack[top--]; // Pop the operator

            bool result;
            if (operator== '&') {
                result = !hasFalse;
            } else if (operator== '|') {
                result = hasTrue;
            } else { // '!'
                result = !hasTrue;
            }

            // Push result back as 't' or 'f'
            stack[++top] = result ? 't' : 'f';
        }
    }

    return stack[0] == 't';
}

File: 1197-parsing-a-boolean-expression/1197-parsing-a-boolean-expression.py
class Solution:
    def parseBoolExpr(self, expression: str) -> bool:
        stack = []
        for curr in expression:
            if curr == ',' or curr == '(':
                continue
            if curr == 't' or curr == 'f' or curr == '!' or curr == '&' or curr == '|':
                stack.append(curr)
            
            elif curr == ')':
                has_true, has_false = False, False
                while stack[-1] != '!' and stack[-1] != '&' and stack[-1] != '|':
                    top_val = stack[-1]
                    stack.pop()
                    if top_val == 't':
                        has_true = True
                    if top_val == 'f':
                        has_false = True
                
                op = stack[-1]
                stack.pop()
                if op == '!':
                    stack.append('f' if has_true else 't')
                elif op == '&':
                    stack.append('f' if has_false else 't')
                else:
                    stack.append('t' if has_true else 'f')
        
        return stack[-1] == 't' 